### PR TITLE
Bundle Deck JavaScript using bazel/rollup

### PR DESCRIPTION
**Pull request by Katharine**
Use bazel, rollup and (a less outdated version of) rules_nodejs to build the deck JavaScript using rollup. (Why rollup? because it's what rules_nodejs supports and absolutely no other reason.)

This is a first step towards using typescript (because in this world it appears we don't use typescript's own bundling).

This PR focuses on making the JavaScript build with minimal changes. It does:

* remove third-party code that was dropped into `static/`
* maks changes where necessary to make `tsc` (used by rules_nodejs for es5 downlevelling) behave properly.

A later PR will focus on actual cleanup.

We also now have bazel manage a single node_modules for the entire repo. Consequently, I have removed the node_modules that previously lived under `triage/`.

This PR still supports `./runlocal` properly and should still generate workable images, but those not using bazel to build their prow images will find that they have no JavaScript.

Also, I have no idea what I'm doing with bazel. Please point out my stupid mistakes.

/area prow/deck
/kind cleanup

### CODE REVIEW

File: prow/cmd/deck/main.go [hunk 1]
```diff
@@ -104,7 +104,7 @@ func gatherOptions() options {
 	flag.StringVar(&o.redirectHTTPTo, "redirect-http-to", "", "Host to redirect http->https to based on x-forwarded-proto == http.")
 	// use when behind an oauth proxy
 	flag.BoolVar(&o.hiddenOnly, "hidden-only", false, "Show only hidden jobs. Useful for serving hidden jobs behind an oauth proxy.")
-	flag.BoolVar(&o.runLocal, "run-local", false, "Serve a local copy of the UI, used by the prow/cmd/deck/runlocal script")
+	flag.StringVar(&o.runLocal, "run-local", "", "Serve a local copy of the UI, used by the prow/cmd/deck/runlocal script")
```
Comment (ixdy) [inline prow/cmd/deck/main.go]: maybe rename this flag and update the description?

File: prow/cmd/deck/runlocal [hunk 2]
```diff
@@ -24,4 +24,4 @@ curl "$HOST/data.js?var=allBuilds" > data.js
 curl "$HOST/tide.js?var=tideData" > tide.js
 curl "$HOST/plugin-help.js?var=allHelp" > plugin-help.js
 curl "$HOST/pr-data.js" > pr-data.js
-bazel run --run_under="cd $DIR && " //prow/cmd/deck:deck -- --run-local --static-files-location=./static --template-files-location=./template --config-path ../../config.yaml
+bazel run //prow/cmd/deck:deck -- --run-local=${DIR}/localdata --static-files-location=./prow/cmd/deck/static --template-files-location=${DIR}/template --config-path ${DIR}/../../config.yaml
```
Comment (ixdy) [inline prow/cmd/deck/runlocal]: in recent versions of Bazel, `bazel run` sets `BUILD_WORKSPACE_DIRECTORY` and `BUILD_WORKING_DIRECTORY`, though I'm not sure that's relevant here.

I guess I'm a little confused why everything except `--static-files-location` is using `${DIR}` though.

Comment (Katharine) [reply prow/cmd/deck/runlocal]: In the case of `--template-files-location`, there is no good reason; I'll change it.

For `--run-local` (which I agree I should rename), the pseudo-static content isn't pulled in by bazel and I see no particular reason to do so. It references $DIR because that's where we happen to put the files (it could just as well be /tmp, and perhaps it should be now that we have dropped the requirement that they live among the static files).

Similarly, for `--config-path`, the config file isn't part of the build, and I don't think it should be.

(`--static-files-location` _must_ refer to the bazel output, which is why it is the only one to do so.)

**Comment by Katharine** [comment]
I am looking at different solutions to node_modules...

**Comment by Katharine** [comment]
@ixdy I've changed the approach to node_modules:

- All of test-infra uses the same node_modules
- The required modules are installed by bazel at build time

This also entailed the removal of the node_modules living in `triage` for consistency.

I'm not sure if this is a better way? I personally like it more. A hybrid approach is possible but I'm not convinced it's a good idea.

**Comment by Katharine** [comment]
Rebased to remove the commit that added a pile of `node_modules`. This should be good to go now.

GitHub's UI does not agree with git on the order of the commits, which is a bit odd. ¯\\\_(ツ)\_/¯

**Comment by stevekuznetsov** [comment]
How does this impact people who build without Bazel? :)

**Comment by Katharine** [comment]
In short, they are now required to do so.

It is probably possible to hack up the `Dockerfile` or otherwise write some script to arrange for an image to be correctly built, but I'm really not a fan of maintaining parallel build systems.

**Comment by stevekuznetsov** [comment]
Hard no on that one, sorry -- breaking downstream users of this by requiring a specific build tool is a no-go as far as I'm concerned.

**Comment by BenTheElder** [comment]
> Hard no on that one, sorry -- breaking downstream users of this by requiring a specific build tool is a no-go as far as I'm concerned.

Discussing this offline ... We can make something work with a Makefile, but "requiring a specific build tool" is always true :-) -- This requires javascript tooling now, which can either be automanaged by bazel, or can be required and wrapped in some make scripts. Wrapping them in make scripts will not change that we have new build dependencies. We're not exactly going to toss node and rollup in vendor/ ...

**Comment by stevekuznetsov** [comment]
> but "requiring a specific build tool" is always true :-)

OK, true, let me rephrase that to "requiring a new build tool and requiring all downstream consumers to migrate immediately"!

**Comment by Katharine** [comment]
@BenTheElder It is possible to have all the JavaScript-y parts managed by yarn, with the exception of yarn and node themselves (which bazel manages, but for another solution would need to be installed at the correct version on the build system).

My primary concern is creating and especially maintaining two parallel sets of configuration: the bazel set, and then a matching relatively complex set of scripts and configuration files for the manual invocation set.

**Comment by BenTheElder** [comment]
Side note: all for not requiring tools where possible. `kind` only uses go and docker and vendors everything possible, builds with go or tools written in go and uses small bits of bash to wrap vendor. test-infra is another beast :-)

I'm excited to see our web tooling refreshed.

> My primary concern is creating and especially maintaining two parallel sets of configuration: the bazel set, and then a matching relatively complex set of scripts and configuration files for the manual invocation set.

Can we maybe verify that the yarn and node versions at least ~ match what we pin in WORKSPACE by matching a --version output? We won't have the same fidelity as pinning specific tool hashes but we can hopefully make it mostly work...

**Comment by ixdy** [comment]
When only using a single language, the language's tools are often sufficient. Where Bazel starts to shine is when you're attempting to integrate multiple tools and languages. For test-infra, that includes Go, Python, Node, and Docker.

It's possible to write our own scripts to manage these integrations, but why should we do that when there are already communities maintaining them for us?

Would it be helpful to discuss this in a Hangout?

**Comment by stevekuznetsov** [comment]
I don't personally have the bandwidth to entirely re-engineer the build and release pipeline for our Prow components. If there's a simple way for someone to help me not be broken once this merges, I'd appreciate it. I'm not in love with container image builds using `Dockerfiles` and whatever tools, but `bazel` is a mystery to me and while I can schedule builds to my k8s cluster right now and manage them just like any other resource, I have no idea how I'd even go about implementing the `bazel` builds.

**Comment by stevekuznetsov** [comment]
If there's a one-stop-shop for `bazel` driven builds I am OK to use that, too, but AFAICT it's complicated today?

**Comment by fejta** [comment]
What do you mean by one stop shop?

**Comment by stevekuznetsov** [comment]
Let's make it easier to use than the `velodrome` deployment :)

**Comment by BenTheElder** [comment]
🙏velodrome is the worst deployment ever, make that better too :-)

**Comment by fejta** [comment]
It should be pretty user-friendly!

```shell
bazel run //prow:release-push
# edit all the cluster/*.yaml files to use the new tag
bazel run //prow/cluster:production.apply
```

And `prow/bump.sh` does all this for you so its just
```shell
prow/bump.sh
bazel run //prow/cluster:production.apply
```

**Comment by fejta** [comment]
You'd need to change these:
https://github.com/kubernetes/test-infra/blob/1ccb6478ed99921a253c4bb432fd80bcc96a2adf/hack/print-workspace-status.sh#L26-L28
to make it push to your registry and kube context

**Comment by stevekuznetsov** [comment]
How do I configure which registry I push to? What does "production apply" do?

**Comment by BenTheElder** [comment]
> How do I configure which registry I push to? What does "production apply" do?

Should be overriding the env (or patching that script to be overridable if it isn't already?)

> What does "production apply" do?

~ kubectl apply of the prow components. note that you should not actually use this as it uses our prow configs unless you've patched those !!

(it's a little bit more than that, rules_k8s does some nice things like resolving images to sha256 references of the images you built from the source tree and those are cached, it replaces the tags for matching images in your manifests with the sha256 to pin them reproducibly) https://github.com/bazelbuild/rules_k8s#apply

**Comment by fejta** [comment]
`export PROW_REPO_OVERRIDE=my/awesome/registry/orefix` should work


`bazel run //whatever:target.verb` will resolve all the yaml files and then run them through `kubectl verb -f`

**Comment by fejta** [comment]
In our case for prod we want stability so don't use some of the more exotic functionality (like dynamically overriding all the tags to an image that this also builds and pushes). 

There's a little but if this in some of the dev jobs, for example https://github.com/kubernetes/test-infra/blob/master/prow/cmd/build/BUILD.bazel#L11 which will build and push whatever is in your client

**Comment by Katharine** [comment]
So, what needs to be done to actually get this in?
